### ISSUE TITLE
Sketch out status-waku module

### ISSUE THREAD

**Issue by oskarth** [status-im/nim-status#12]
Currently status-go implements Waku v1 in Go (see here https://github.com/status-im/status-go/blob/develop/waku)

As part of nim-status, we should
  - include [nim-waku v1](https://github.com/status-im/nim-waku/tree/master/waku/node/v1) as a module and re-use it instead of native status-go code. Follow RPC spec defined [here](https://github.com/vacp2p/specs/blob/master/specs/waku/waku-rpc-api.md)
  - in parallel, also provide API endpoints for [waku v2](https://github.com/status-im/nim-waku/tree/master/waku/node/v2). We should also provide a v1-v2 switching capability

Points for clarification:
  - for RPC calls, the node will default to the highest Waku version available. Otherwise, we need to define the version in the JSON-RPC payload. Which option is better?
  - nim-waku support for extensions

**Comment by oskarth** [comment]
@siphiuel @siphiuel created a skeleton issue, please fill out with details as you see fit and we can discuss on Monday

**Comment by oskarth** [comment]
@siphiuel @Samyoul (lol did double mention before) - any luck with this issue? Is there some scratch work somewhere?

**Comment by andremedeiros** [comment]
> Is there some scratch work somewhere?

I'd also like to see at least an issue where progress is being tracked!

**Comment by oskarth** [comment]
@andremedeiros This is the issue.

@iurimatias why did you unassign them ? This is part of the Waku 2 work.

**Comment by siphiuel** [comment]
Just started working on it (importing some nim-waku functions into nim-status), will push a draft PR soon.

**Comment by oskarth** [comment]
@Samyoul @siphiuel If this issue is too big, can we split it up so it can be finished on the timeframe of a few days please? Also if it is in progress mark it as such in https://github.com/vacp2p/research/projects/1?add_cards_query=is%3Aopen

**Comment by siphiuel** [comment]
An outcome of the  discussion in https://notes.status.im/-94bFPyzSw2JM_FIXMij3Q is this plan that we'd like to discuss:
  - create an API endpoint in front of nim-waku/status-go
  - handle `waku_*` by nim-waku
  - handle `wakuext_*` by status-go
  - thus there will 2 Ethereum client instances started: one owned by nim-waku and another by status-go

Issues as pointed out by @cammellos:
  - some methods e.g. `waku_markTrustedPeer` will not properly work without additional communication mechanisms between Nim and Go
  - a suggestion is to `Change status-go code so that any interaction with waku is top level, i.e. surface the data in nim, and use nim-waku to handle the data to be dispatched. Then pass to status-go for the processing, and pass finally to the client.` @cammellos can you please explain this a bit more?

**Comment by cammellos** [comment]
So,
say for example you have two instances running of `waku`. One in `status-go` and the other one in `nim`.

```
   "waku_generateSymKeyFromPassword" {}
   "waku_getSymKey" {}
   "waku_markTrustedPeer" {}
```

The first two method, it would kind of work, as we really care only about the actual sym key and not the id.
But let's take `markTrustedPeer` as an example.

`markTrustedPeer` is used to notify `waku` that it should allow historical messages from this peer

First it's unclear what the benefit would be. If I receive a message on `waku_nim` instance, where would this message go? likely it will be just ignored, as status-react does not hit any `waku_` endpoint to install filters etc, so it would just be running but not receiving any message.

But basically say that you have these two instances running, and you hit `markTrustedPeer` on the nim-waku instance.

Immediately you run in a bunch of issues:

1) Peers are not handled by waku, peers are handled by the geth node. We call `addPeer` on the bindings that goes through geth, so you would have to `addPeer` on nim-waku as well (I guess that's ok to do).

2) At this point, you will have a trusted peer in nim-waku (which will not receive messages really, as all the filters are installed on status-go nim), and a non-trusted peer in status-go.When we receive messages from this peer in status-go, messages will be discarded in status-go, as it's not trusted. So effectively you will have to run `markTrustedPeer` on status-go as well.

Essentially having two instances running does not get you any closer to integration, as you won't be really using nim-waku, the only thing is actually doing (which might be valuable), is "can we run nim-waku in status-react/desktop".

The main issue is that 
`wakuext_....` endpoints, heavily use `waku_...`.

The solution that we proposed is to shim nim -> status-go, but status-go uses waku underneath. So if you really want to keep nim->status-go and not allow nim->status-go->nim, or status-go->nim, the only way to do this is something like:
assume this is nim code:
```
func sendChatMessage(blah) {
  encryptedChatMessage := statusgo.buildEncryptChatMessage(blah)
  nim-waku.dispatch(encryptedChathMessage
}
```
Now what is really happening is (this is go code):

```
func sendChatMessage(blah) {
    buildProtobufMessage(blah)
    passToDatasync()  // asynchrous, datasync has it's own tick 
}

func datasyncTick() {
   // check for messages to be dispatched
  encryptMessage(msg)
  dispatchWithWaku(msg)
}
```

So you can see that is fairly complex. 
Pedro's approach was to pass a C library of waku-nim to status-go, in which case you don't have to touch existing code much, and you can just replace nim-waku at the lower level, instead of having to surface data to be dispatched in nim

**Comment by oskarth** [comment]
Does closing it means it has been done? Would be good to have a comment summarizing current state, reason for closing, or pointer to better scoped issue.
